### PR TITLE
rockstar+RAMSES functionality

### PR DESCRIPTION
**Pull request by AnatoleStorck**
Added functionality for adding RAMSES datasets with rockstar-galaxies catalogues using yt to tangos.

### CODE REVIEW

**Comment by apontzen** [comment]
Thanks @AnatoleStorck @cphyc  - do you understand why the enzo test database now fails to build?

If needed you can try this locally by running `export INTEGRATION_TESTING=1; ./build.sh` in the `test_tutorial_build` folder

**Comment by AnatoleStorck** [comment]
I have moved the new functions in the general YT handler which were causing the Enzo test to fail (I still do not understand why) into the input handler for RAMSES+rockstar. The Enzo test now passes. Since this pull request is for supporting RAMSES+rockstar functionality, this should be fine.

**Comment by apontzen** [comment]
To me it looks like people who have developed their own input handlers that derive from `YtInputHandler` will be tripped up by this. Unfortunately I think it's necessary to solve the underlying problem rather than remove functionality from the crucial base class.

(I am not sure why the tests are not actually running on Github Actions, by the way - will see if I can figure that out)

**Comment by AnatoleStorck** [comment]
Fixed the issue, occurring due to the handler not importing the statfile if it exists.

File: tangos/input_handlers/yt.py [hunk 1]
```diff
@@ -85,6 +170,130 @@ def _load_halo_cat_without_caching(self, ts_extension, snapshot_file):
...
+            overdir = self._extension_to_filename("")
+            snapfiles = glob.glob(overdir+ts_extension[:2]+len(ts_extension[2:].split('/')[0])*'?')
+            rockfiles = glob.glob(overdir+"out_*.list")
+            sortind = np.array([int(rname.split('.')[0].split('_')[-1]) for rname in rockfiles])
+            sortord = np.argsort(sortind)
+            snapfiles.sort()
+            rockfiles = np.array(rockfiles)[sortord]
+            timestep_ind = np.argwhere(np.array([s.split('/')[-1] for s in snapfiles])==ts_extension.split('/')[0])[0]
+            fnum = int(rockfiles[timestep_ind][0].split('.')[0].split('_')[-1])
+        cat = yt.load(self._extension_to_filename("halos_"+str(fnum)+".0.bin"))
```
Comment (cphyc) [inline tangos/input_handlers/yt.py:250]:
```suggestion
        cat = yt.load(self._extension_to_filename(f"halos_{fnum}.0.bin"))
```

File: tangos/input_handlers/yt.py [hunk 2]
```diff
@@ -85,6 +170,130 @@ def _load_halo_cat_without_caching(self, ts_extension, snapshot_file):
...
+        cat_data = cat.all_data()
+        # Check whether rockstar was run with Behroozi's distribution or Wise's
+        if np.any(cat_data["halos","particle_identifier"]<0):
+            del cat
+            del cat_data
+            cat = yt.load(self._extension_to_filename("halos_"+str(fnum)+".0.bin"))
```
Comment (cphyc) [inline tangos/input_handlers/yt.py:256]:
```suggestion
            cat = yt.load(self._extension_to_filename(f"halos_{fnum}.0.bin"))
```

File: tangos/input_handlers/yt.py [hunk 3]
```diff
@@ -85,6 +170,130 @@ def _load_halo_cat_without_caching(self, ts_extension, snapshot_file):
...
+            return filter
+
+        add_particle_filter("mrp_dark_matter", function=MustRefineParticles, filtered_type='all', \
+                            requires=["particle_family"])
+
+        f.add_particle_filter("stars")
+        f.add_particle_filter("dark_matter")
+        f.add_particle_filter("mrp_dark_matter")
+
+        return f
```
Comment (cphyc) [inline tangos/input_handlers/yt.py:234]: For the past 6 years, yt automatically creates those filters if the `particle_family` array already exists, so those lines are actually mostly redundant (and should be yt-specific anyways). I'd suggest
```suggestion
        if mode is not None:
            raise ValueError("Custom load modes are not supported with yt")
        return yt.load(self._extension_to_filename(ts_extension))
```

File: tangos/input_handlers/yt.py [hunk 4]
```diff
@@ -85,6 +170,130 @@ def _load_halo_cat_without_caching(self, ts_extension, snapshot_file):
...
+            snapfiles.sort()
+            rockfiles = np.array(rockfiles)[sortord]
+            timestep_ind = np.argwhere(np.array([s.split('/')[-1] for s in snapfiles])==ts_extension.split('/')[0])[0]
+            fnum = int(rockfiles[timestep_ind][0].split('.')[0].split('_')[-1])
+        cat = yt.load(self._extension_to_filename("halos_"+str(fnum)+".0.bin"))
+        cat_data = cat.all_data()
+        # Check whether rockstar was run with Behroozi's distribution or Wise's
+        if np.any(cat_data["halos","particle_identifier"]<0):
+            del cat
+            del cat_data
```
Comment (cphyc) [inline tangos/input_handlers/yt.py:255]: This is useless since we overwrite those below anyway.
```suggestion
```

File: tangos/input_handlers/yt.py [hunk 5]
```diff
@@ -85,6 +170,130 @@ def _load_halo_cat_without_caching(self, ts_extension, snapshot_file):
...
+            for i in range(num_objects):
+                obj = self.load_object(
+                    ts_extension,
+                    int(catalogue_data["halos","particle_identifier"][i]),
+                    i,
+                    object_typetag
+                )
+                NDM = len(obj["dark_matter","particle_mass"])
+                NGas = 0 # cells
+                NStar = len(obj["stars","particle_mass"])
```
Comment (cphyc) [inline tangos/input_handlers/yt.py:285]: If you accept my suggested change above, this should become
```suggestion
                NDM = len(obj["DM", "particle_ones"])
                NGas = 0 # cells
                NStar = len(obj["star", "particle_ones"])
```
Note that I'm using `particle_ones` rather than `particle_mass` so that we don't have to read anything from the disk.

File: tangos/input_handlers/yt.py [hunk 6]
```diff
@@ -85,6 +170,130 @@ def _load_halo_cat_without_caching(self, ts_extension, snapshot_file):
...
+                if NDM + NGas + NStar> min_halo_particles:
+                    yield i, int(catalogue_data["halos","particle_identifier"][i]), NDM, NStar, NGas
+
+    def load_object(self, ts_extension, finder_id, finder_offset, object_typetag='halo', mode=None):
+        f = self.load_timestep(ts_extension, mode)
+        cat, cat_dat = self._load_halo_cat(ts_extension, object_typetag)
+        center = cat_dat["halos","particle_position"][cat_dat["halos","particle_identifier"]==finder_id][0]
+        center+=f.domain_left_edge-cat.domain_left_edge
+        radius = cat_dat["halos","virial_radius"][cat_dat["halos","particle_identifier"]==finder_id][0]
+        return f.sphere(center.in_cgs(), radius.in_cgs())
```
Comment (cphyc) [inline tangos/input_handlers/yt.py:295]:
```suggestion
        f = self.load_timestep(ts_extension, mode)
        cat, cat_dat = self._load_halo_cat(ts_extension, object_typetag)
        index = np.argwhere(cat_dat["halos", "particle_identifier"] == finder_id)[0, 0]
        center = cat_dat["halos","particle_position"][index]
        center += f.domain_left_edge - cat.domain_left_edge
        radius = cat_dat["halos", "virial_radius"][index]
        return f.sphere(center, radius)
```

File: tangos/tools/add_simulation.py [hunk 7]
```diff
@@ -150,7 +150,8 @@ def add_objects_to_timestep(self, ts, create_class=core.halo.Halo):
                 halos.append(h)
 
         with pt.ExclusiveLock("db_write_lock"):
-            logger.info("Add %d %ss to timestep %r", len(halos), create_class.__name__, ts)
+            if (create_class.__name__ != "Group"):
```
Comment (cphyc) [inline tangos/tools/add_simulation.py:153]: What's the motivation for this change?

```suggestion
            if create_class.__name__ != "Group":
```

File: tangos/input_handlers/yt.py [hunk 8]
```diff
@@ -85,6 +170,130 @@ def _load_halo_cat_without_caching(self, ts_extension, snapshot_file):
...
+        return cat, cat_data
+
+    def enumerate_objects(self, ts_extension, object_typetag="halo", min_halo_particles=config.min_halo_particles):
+        if object_typetag!="halo":
+            return
+        if self._can_enumerate_objects_from_statfile(ts_extension, object_typetag):
+            yield from self._enumerate_objects_from_statfile(ts_extension, object_typetag)
+        else:
+            #logger.warn("No halo statistics file found for timestep %r", ts_extension)
+            #logger.warn(" => enumerating %ss directly using yt", object_typetag)
```
Comment (cphyc) [inline tangos/input_handlers/yt.py:268]: Is there a reason to comment that out?

File: tangos/input_handlers/yt.py [hunk 9]
```diff
@@ -85,6 +170,130 @@ def _load_halo_cat_without_caching(self, ts_extension, snapshot_file):
...
+            cat = yt.load(self._extension_to_filename("halos_"+str(fnum)+".0.bin"))
+            cat.parameters['format_revision'] = 2 #
+            cat_data = cat.all_data()
+        return cat, cat_data
+
+    def enumerate_objects(self, ts_extension, object_typetag="halo", min_halo_particles=config.min_halo_particles):
+        if object_typetag!="halo":
+            return
+        if self._can_enumerate_objects_from_statfile(ts_extension, object_typetag):
+            yield from self._enumerate_objects_from_statfile(ts_extension, object_typetag)
```
Comment (cphyc) [inline tangos/input_handlers/yt.py:265]: I'd suggest returning after the yield, something like:
```
            yield from self._enumerate_objects_from_statfile(ts_extension, object_typetag)
            return
```
This way you can de-indent the rest of the function which is always nice :)

File: tangos/input_handlers/yt.py [hunk 10]
```diff
@@ -85,6 +170,130 @@ def _load_halo_cat_without_caching(self, ts_extension, snapshot_file):
...
+        if os.path.exists(self._extension_to_filename("datasets.txt")):
+            fnum = read_datasets(self._extension_to_filename(""),ts_extension)
+        else: # otherwise, assume a one-to-one correspondence
+            overdir = self._extension_to_filename("")
+            snapfiles = glob.glob(overdir+ts_extension[:2]+len(ts_extension[2:].split('/')[0])*'?')
+            rockfiles = glob.glob(overdir+"out_*.list")
+            sortind = np.array([int(rname.split('.')[0].split('_')[-1]) for rname in rockfiles])
+            sortord = np.argsort(sortind)
+            snapfiles.sort()
+            rockfiles = np.array(rockfiles)[sortord]
```
Comment (cphyc) [inline tangos/input_handlers/yt.py:247]: Do you know where this piece of code comes from? It's really hard to make sense out of it and I'm wondering if we could simplify it somehow.

File: tangos/input_handlers/yt.py [hunk 11]
```diff
@@ -46,8 +47,92 @@ def load_object(self, ts_extension, finder_id, finder_offset, object_typetag='ha
...
+            if not (halo_min <= ihalo1_tangos <= halo_max):
+                continue
+
+            ids1 = h1.halo("halos", ihalo1_rockstar).member_ids
+            #mask = np.in1d(ids1, members2)
+            mask = np.in1d(members2, ids1)
+            if mask.sum() == 0:
+                cat.append(
+                    []
+                )
```
Comment (cphyc) [inline tangos/input_handlers/yt.py:107]:
```suggestion
                cat.append([])
```

**Review by cphyc — COMMENTED** [review]
Looks great to me! I've left extensive comments to make it more yt-esque!

**Comment by apontzen** [comment]
If you can pull master into this branch, it will also run the unit tests. At the moment only the integration test is running due to a configuration error in github actions.

File: tangos/input_handlers/yt.py (hunk 10 above)
Comment (AnatoleStorck) [reply tangos/input_handlers/yt.py:247]: So the YtRamsesRockstarInputHandler was first created by taking the YtEnzoRockstarInputHandler and adapting it to ramses datasets. This function is taken from there.

File: tangos/tools/add_simulation.py (hunk 7 above)
Comment (AnatoleStorck) [reply tangos/tools/add_simulation.py:153]: This change should be removed, was put there for debugging purposes.

**Review by AnatoleStorck — COMMENTED** [review]
Thank you for the suggestions! I will also remove the logging changes
